fastlane/fastlane
# [CI] improve bundler's cache on CircleCI

### Checklist

### Motivation and Context



Note: There is another my PR #17861 that conflicts to this PR. Either way, I will resolve the conflict after one gets merged🙇 

----

While working on #17861 and #17863, I realised that bundler cache doesn't work. This is because the cache key is just wrong from the beginning to be used across all Ruby environments.

In this PR, I fixed the cache-key to make caching work properly. To do so, I also migrated to `.circleci/config.yml` to the latest version "2.1" to describe parameterised builds.

If the cache works correctly, build time would be shortened by 1-2 mins or so presumably... 

### Description



> v2-gems-{{ arch }}-{{ .Branch }}-{{ checksum "Gemfile" }}-{{ checksum "fastlane.gemspec" }}

This is the cache key to store/restore installed gems. This is just wrong in the following reasons. 

* The cache (un)intentionally will be used from all the Ruby version as long as "arch" is the same
    * `bundler` keep installed gems separated by Ruby version, which means the cache works only for one of Ruby versions tested
* The cache will be available only within the same branch's builds
* Appending checksums of "Gemfile" and "gemspec" isn't accurate enough to serve all the required dependencies precisely 
   * Eventually `bundler` resolves dependencies but it's not 100% efficient (cache miss)

(FYI: `arch` matters if we use a gem that has native extensions.)

I'm going to change the key to this. 

> v3-gems-{{ arch }}-<< parameters.ruby_version >>-{{ checksum "Gemfile.lock" }}

This way 

- The cache will be kept by each Ruby version or docker image
- The cache will be shared across all the branches
- Builds can pick up exactly correct cache according to `Gemfile.lock`

To do this, I had to switch CircleCI's config format version to "2.1" as well. It allows you to reuse a "job" definition and run it against different configurations using parameters (a.k.a parameterised builds). See https://circleci.com/docs/2.0/reusing-config/

### Testing Steps



See if "bundle install" step is done within a few seconds across all the Ruby version/environments to install "bundler" gem.

<img width="868" alt="Screenshot 2021-01-03 at 00 17 52" src="https://user-images.githubusercontent.com/748949/103469163-2732c980-4d59-11eb-81e9-e9498faed15f.png">

https://app.circleci.com/pipelines/github/fastlane/fastlane/1899/workflows/b9988e82-2674-4ba4-abe1-096a051300a3

In the above build, the following links are the corresponding "bundle install" step's logs.

* Ruby 2.6 https://app.circleci.com/pipelines/github/fastlane/fastlane/1899/workflows/b9988e82-2674-4ba4-abe1-096a051300a3/jobs/59652/parallel-runs/0/steps/0-109
* Ruby 2.5 https://app.circleci.com/pipelines/github/fastlane/fastlane/1899/workflows/b9988e82-2674-4ba4-abe1-096a051300a3/jobs/59651/parallel-runs/0/steps/0-109
* Ruby 2.4 https://app.circleci.com/pipelines/github/fastlane/fastlane/1899/workflows/b9988e82-2674-4ba4-abe1-096a051300a3/jobs/59658/parallel-runs/0/steps/0-109
* Ubuntu https://app.circleci.com/pipelines/github/fastlane/fastlane/1899/workflows/b9988e82-2674-4ba4-abe1-096a051300a3/jobs/59653/parallel-runs/0/steps/0-106
